datastore/elastic.go

```diff
@@ -0,0 +1,181 @@
+package datastore
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/bar41234/bar_book_service/models"
+	"github.com/olivere/elastic/v7"
+	"strconv"
+	"strings"
+)
+
+const (
+	ElasticUrl = "http://es-search-7.fiverrdev.com:9200"
+	IndexName  = "bar_books_store_2"
+	QuerySize  = 100
+)
+
+type ElasticDb struct {
+	client *elastic.Client
+	isInit bool
+}
+
+func (e *ElasticDb) Init() error {
+	if !e.isInit {
+		client, err := elastic.NewClient(elastic.SetURL(ElasticUrl))
+		if err != nil {
+			return err
+		}
+		e.client = client
+		ctx := context.Background()
+		_, _, err = e.client.Ping(ElasticUrl).Do(ctx)
+		if err != nil {
+			return err
+		}
+		e.isInit = true
+	}
+	return nil
+}
+
+func (e *ElasticDb) Get(id string) (*models.Book, error) {
+	initErr := e.Init()
+	if initErr != nil {
+		return nil, initErr
+	}
+	ctx := context.Background()
+	get, getErr := e.client.Get().Index(IndexName).Id(id).Do(ctx)
+	if getErr != nil {
+		return nil, getErr
+	}
+	var book models.Book
+	unmarshalErr := json.Unmarshal(get.Source, &book)
+	if unmarshalErr != nil {
+		return nil, unmarshalErr
+	}
+	return &book, nil
+}
+
+func (e *ElasticDb) Put(book models.Book) (string, error) {
+	initErr := e.Init()
+	if initErr != nil {
+		return "", initErr
+	}
+	ctx := context.Background()
+	bookJson, marshalErr := json.Marshal(book)
+	if marshalErr != nil {
+		return "", marshalErr
+	}
+	bj := string(bookJson)
+	put, putErr := e.client.Index().
+		Index(IndexName).
+		BodyJson(bj).
+		Do(ctx)
+	if putErr != nil {
+		return "", putErr
+	}
+	return put.Id, nil
+
+}
+
+func (e *ElasticDb) Post(shortBook models.ShortBook) (*models.Book, error) {
+	initErr := e.Init()
+	if initErr != nil {
+		return nil, initErr
+	}
+	ctx := context.Background()
+	_, updateErr := e.client.Update().Index(IndexName).Id(shortBook.Id).Doc(map[string]interface{}{"title": shortBook.Title}).Do(ctx)
+	if updateErr != nil {
+		return nil, updateErr
+	}
+	return e.Get(shortBook.Id)
+}
+
+func (e *ElasticDb) Delete(id string) error {
+	initErr := e.Init()
+	if initErr != nil {
+		return initErr
+	}
+	ctx := context.Background()
+	_, deleteErr := e.client.Delete().Index(IndexName).Id(id).Do(ctx)
+	return deleteErr
+}
+
+func (e *ElasticDb) queryBuilder(bookQuery models.BookQuery) (*elastic.BoolQuery, error) {
+	query := elastic.NewBoolQuery()
+	if bookQuery.Title != "" {
+		titleQuery := elastic.NewMatchQuery("title", bookQuery.Title)
+		query.Must(titleQuery)
+	}
+	if bookQuery.AuthorName != "" {
+		authorQuery := elastic.NewMatchQuery("author_name", bookQuery.AuthorName)
```

Comment:
why do you use match query and not tearm query?

Reply:
Maybe the user searched for first name without family name? (for example, author name 'Mark' instead of 'Mark Twain')

Reply:
did you index `author_name` field as a text field?

Reply:
yep